core/trino-main/src/main/java/io/trino/execution/scheduler/EventDrivenFaultTolerantQueryScheduler.java

```diff
@@ -748,32 +798,164 @@ private void updateStageExecutions()
             });
         }
 
-        private boolean isReadyForExecution(SubPlan subPlan)
+        private static class IsReadyForExecutionResult
         {
-            for (SubPlan child : subPlan.getChildren()) {
-                StageExecution childExecution = stageExecutions.get(getStageId(child.getFragment().getId()));
-                if (childExecution == null) {
-                    return false;
+            private final boolean readyForExecution;
+            private final Optional<Map<StageId, OutputDataSizeEstimate>> sourceOutputSizeEstimates;
+
+            @CheckReturnValue
+            public static IsReadyForExecutionResult ready(Map<StageId, OutputDataSizeEstimate> sourceOutputSizeEstimates)
+            {
+                return new IsReadyForExecutionResult(true, Optional.of(sourceOutputSizeEstimates));
+            }
+
+            @CheckReturnValue
+            public static IsReadyForExecutionResult notReady()
+            {
+                return new IsReadyForExecutionResult(false, Optional.empty());
+            }
+
+            private IsReadyForExecutionResult(boolean readyForExecution, Optional<Map<StageId, OutputDataSizeEstimate>> sourceOutputSizeEstimates)
+            {
+                requireNonNull(sourceOutputSizeEstimates, "sourceOutputSizeEstimates is null");
+                if (readyForExecution) {
+                    checkArgument(sourceOutputSizeEstimates.isPresent(), "expected sourceOutputSizeEstimates to be set");
                 }
-                // TODO enable speculative execution
-                if (childExecution.getState() != StageState.FINISHED) {
-                    return false;
+                if (!readyForExecution) {
+                    checkArgument(sourceOutputSizeEstimates.isEmpty(), "expected sourceOutputSizeEstimates to be not set");
                 }
+                this.readyForExecution = readyForExecution;
+                this.sourceOutputSizeEstimates = sourceOutputSizeEstimates.map(ImmutableMap::copyOf);
             }
-            return true;
+
+            public boolean isReadyForExecution()
+            {
+                return readyForExecution;
+            }
+
+            public Map<StageId, OutputDataSizeEstimate> getSourceOutputSizeEstimates()
+            {
+                return sourceOutputSizeEstimates.orElseThrow();
+            }
+        }
+
+        private IsReadyForExecutionResult isReadyForExecution(SubPlan subPlan)
+        {
+            boolean nonSpeculativeTasksInQueue = schedulingQueue.getNonSpeculativeTaskCount() > 0;
+            boolean nonSpeculativeTasksWaitingForNode = preSchedulingTaskContexts.values().stream()
+                    .anyMatch(task -> !task.isSpeculative() && !task.getNodeLease().getNode().isDone());
+
+            // do not start a speculative stage if there is non-speculative work still to be done.
+            boolean canScheduleSpeculative = !nonSpeculativeTasksInQueue && !nonSpeculativeTasksWaitingForNode;
```

Comment:
nit: maybe "should"?

Reply:
I will leave it as is. I think `can` conveys the intent slightly better